### PR TITLE
drivers/video: add support for V4L2 mmap-ed buffer

### PR DESCRIPTION
**Pull request by PeterBee97**
## Summary
Added a naive implementation for V4L2_MEMORY_MMAP type buffers.
## Impact
Enhancement to NuttX video drivers

### CODE REVIEW

File: drivers/video/video.c [hunk 1]
```diff
@@ -1161,6 +1162,15 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
 
   memcpy(&container->buf, buf, sizeof(struct v4l2_buffer));
   video_framebuff_queue_container(&type_inf->bufinf, container);
+  if (g_video_data_ops->enq_buf)
+    {
+      ret = g_video_data_ops->enq_buf((uint8_t *)(uintptr_t)buf->m.userptr,
+          buf->length);
```
Comment (acassis) [inline drivers/video/video.c]: align buf->length); with the beginning of ( of previous line!

File: video/Kconfig [hunk 2]
```diff
@@ -13,6 +13,10 @@ config VIDEO
 
 if VIDEO
 
+config VIDEO_DEV_PATH
+	string "video device path"
```
Comment (acassis) [inline video/Kconfig:17]: Please use capitalized string: "Video Device Path"

Comment (PeterBee97) [reply video/Kconfig]: Done

File: drivers/video/video.c [hunk 3]
```diff
@@ -62,6 +62,9 @@
 
 #define VIDEO_ID(x, y) (((x) << 16) | y)
 
+#define MAX_VIDEO_HEAP_SIZE CONFIG_VIDEO_HEAP_SIZE
+#define VMEM_HEAP(v,i) ((uint8_t*)v->vmem_heap + (i) * v->buf_size)
```
Comment (pkarashchenko) [inline drivers/video/video.c:66]:
```suggestion
#define VMEM_HEAP(v,i) ((uint8_t *)(v)->vmem_heap + (i) * (v)->buf_size)
```

File: drivers/video/video.c [hunk 4]
```diff
@@ -62,6 +62,9 @@
 
 #define VIDEO_ID(x, y) (((x) << 16) | y)
```
Comment (pkarashchenko) [inline drivers/video/video.c:63]:
```suggestion
#define VIDEO_ID(x, y) (((x) << 16) | (y))
```

File: drivers/video/video.c [hunk 5]
```diff
@@ -134,6 +137,9 @@ struct video_mng_s
   uint8_t            open_num;
   video_type_inf_t   video_inf;
   video_type_inf_t   still_inf;
+  void               *vmem_heap;   /* for V4L2_MEMORY_MMAP buffers */
```
Comment (pkarashchenko) [inline drivers/video/video.c:140]:
```suggestion
  FAR void           *vmem_heap;   /* for V4L2_MEMORY_MMAP buffers */
```

File: drivers/video/video.c [hunk 6]
```diff
@@ -1161,6 +1290,24 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
 
   memcpy(&container->buf, buf, sizeof(struct v4l2_buffer));
   video_framebuff_queue_container(&type_inf->bufinf, container);
+  if (g_video_data_ops->enq_buf)
```
Comment (pkarashchenko) [inline drivers/video/video.c:1293]: Optional
```suggestion
  if (g_video_data_ops->enq_buf != NULL)
```

File: drivers/video/video.c [hunk 7]
```diff
@@ -1161,6 +1290,24 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
...
+    {
+      if (buf->memory == V4L2_MEMORY_MMAP)
+        {
+          ret = g_video_data_ops->enq_buf(
+            VMEM_HEAP(vmng, buf->index), vmng->buf_size);
+        }
+      else
+        {
+          ret = g_video_data_ops->enq_buf(
+            (uint8_t *)(uintptr_t)buf->m.userptr, buf->length);
```
Comment (pkarashchenko) [inline drivers/video/video.c:1303]:
```suggestion
            (FAR uint8_t *)(uintptr_t)buf->m.userptr, buf->length);
```

File: drivers/video/video.c [hunk 8]
```diff
@@ -1468,6 +1631,22 @@ static int validate_frame_setting(enum v4l2_buf_type type,
...
+{
+  size_t ret = vf->width * vf->height;
+  switch (vf->pixelformat)
+    {
+      case V4L2_PIX_FMT_YUV420 :
+        return ret * 3 / 2;
+      case V4L2_PIX_FMT_YUYV :
+      case V4L2_PIX_FMT_UYVY :
+      case V4L2_PIX_FMT_RGB565 :
+      case V4L2_PIX_FMT_JPEG :
```
Comment (pkarashchenko) [inline drivers/video/video.c:1644]:
```suggestion
      case V4L2_PIX_FMT_YUV420:
        return ret * 3 / 2;
      case V4L2_PIX_FMT_YUYV:
      case V4L2_PIX_FMT_UYVY:
      case V4L2_PIX_FMT_RGB565:
      case V4L2_PIX_FMT_JPEG:
```

File: include/nuttx/video/imgdata.h [hunk 9]
```diff
@@ -73,6 +76,8 @@ struct imgdata_ops_s
 
   CODE int (*validate_buf)(uint8_t *addr, uint32_t size);
   CODE int (*set_buf)(uint8_t *addr, uint32_t size);
+  CODE int (*enq_buf)(uint8_t *addr, uint32_t size);
+  CODE int (*dq_buf)(uint8_t **addr, struct timeval *ts);
```
Comment (pkarashchenko) [inline include/nuttx/video/imgdata.h:80]:
```suggestion
  CODE int (*validate_buf)(FAR uint8_t *addr, uint32_t size);
  CODE int (*set_buf)(FAR uint8_t *addr, uint32_t size);
  CODE int (*enq_buf)(FAR uint8_t *addr, uint32_t size);
  CODE int (*dq_buf)(FAR uint8_t **addr, FAR struct timeval *ts);
```

File: include/nuttx/video/video.h [hunk 10]
```diff
@@ -641,6 +1097,89 @@ struct v4l2_querymenu
...
+{
+    u_int32_t         index;       /*  Which input */
+    u_int8_t          name[32];    /*  Label */
+    u_int32_t         type;        /*  Type of input */
+    u_int32_t         audioset;    /*  Associated audios (bitfield) */
+    u_int32_t         tuner;       /*  enum v4l2_tuner_type */
+    v4l2_std_id  std;
+    u_int32_t         status;
+    u_int32_t         capabilities;
+    u_int32_t         reserved[3];
```
Comment (pkarashchenko) [inline include/nuttx/video/video.h:1110]: 2 spaces alignment

File: include/nuttx/video/video.h [hunk 11]
```diff
@@ -305,6 +630,136 @@ struct v4l2_selection
...
+                 V4L2_STD_625_50)
+
+struct v4l2_standard
+{
+    u_int32_t             index;
+    v4l2_std_id           id;
+    u_int8_t              name[24];
+    struct v4l2_fract     frameperiod; /* Frames, not fields */
+    u_int32_t             framelines;
+    u_int32_t             reserved[4];
```
Comment (pkarashchenko) [inline include/nuttx/video/video.h:760]: 2 spaces alignement.
Also why not using stdint.h types?

File: include/nuttx/video/video.h (hunk 10 above)
Comment (PeterBee97) [reply include/nuttx/video/video.h]: done

File: include/nuttx/video/imgdata.h (hunk 9 above)
Comment (PeterBee97) [reply include/nuttx/video/imgdata.h]: fixed

File: drivers/video/video.c (hunk 8 above)
Comment (PeterBee97) [reply drivers/video/video.c:1634]: fixed

File: drivers/video/video.c (hunk 6 above)
Comment (PeterBee97) [reply drivers/video/video.c]: done

File: drivers/video/video.c (hunk 5 above)
Comment (PeterBee97) [reply drivers/video/video.c]: fixed

**Comment by xiaoxiang781216** [comment]
@PeterBee97 please rebase this PR.

File: drivers/video/video.c [hunk 12]
```diff
@@ -978,6 +985,11 @@ static void cleanup_resources(FAR video_mng_t *vmng)
   cleanup_streamresources(&vmng->video_inf);
   cleanup_streamresources(&vmng->still_inf);
   cleanup_scenes_parameter();
+  if (vmng->vmem_heap != NULL)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:988]: move to cleanup_streamresources

File: drivers/video/video.c [hunk 13]
```diff
@@ -1172,6 +1184,16 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
     }
   else
     {
+      if (reqbufs->memory == V4L2_MEMORY_MMAP)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1187]: remove the check since the config should work for non-map mode too.

File: drivers/video/video.c [hunk 14]
```diff
@@ -1180,9 +1202,51 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
 
   leave_critical_section(flags);
 
+  if ((ret == OK) && (reqbufs->memory == V4L2_MEMORY_MMAP))
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1205]: let's move into else clause and before line 1197

File: drivers/video/video.c [hunk 15]
```diff
@@ -1180,9 +1202,51 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
...
+    {
+      vmng->buf_count = reqbufs->count;
+      if (vmng->vmem_heap != NULL)
+        {
+          kmm_free(vmng->vmem_heap);
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1210]: let's use kmm_realloc instead kmm_free/kmm_malloc

File: drivers/video/video.c [hunk 16]
```diff
@@ -1180,9 +1202,51 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
...
+      return -EINVAL;
+    }
+
+  if (buf->index > vmng->buf_count - 1)
+    {
+      return -EINVAL;
+    }
+
+  buf->m.offset = vmng->buf_size * buf->index;
+  buf->length = vmng->buf_size;
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1245]: should you attach buffer before enqueue at line 1280

File: drivers/video/video.c (hunk 15 above)
Comment (PeterBee97) [reply drivers/video/video.c:1210]: Can realloc ensure alignment of address similarly to memalign?

File: drivers/video/video.c [hunk 17]
```diff
@@ -219,6 +223,7 @@ static int validate_frame_setting(enum v4l2_buf_type type,
                                   uint8_t nr_fmt,
                                   FAR video_format_t *vfmt,
                                   FAR struct v4l2_fract *interval);
+static size_t video_fmt_buf_size(video_format_t *vf);
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:226]:
```suggestion
static size_t video_fmt_buf_size(FAR video_format_t *vf);
```

File: drivers/video/video.c [hunk 18]
```diff
@@ -1213,6 +1282,8 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
       return -ENOMEM;
     }
 
+  buf->length = video_fmt_buf_size(&type_inf->fmt[VIDEO_FMT_MAIN]);
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1285]: need check buf->memory == V4L2_MEMORY_MMAP before assignment

File: drivers/video/video.c [hunk 19]
```diff
@@ -1213,6 +1282,8 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
...
+  buf->m.offset = buf->length * buf->index;
   memcpy(&container->buf, buf, sizeof(struct v4l2_buffer));
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1287]: let's update container->buf instead buf

File: drivers/video/video.c [hunk 20]
```diff
@@ -1236,13 +1307,27 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
...
           if (container)
             {
-              start_capture(buf->type,
-                            type_inf->nr_fmt,
-                            type_inf->fmt,
-                            &type_inf->clip,
-                            &type_inf->frame_interval,
-                            container->buf.m.userptr,
-                            container->buf.length);
+              if (container->buf.memory == V4L2_MEMORY_MMAP)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1310]: why do you need check V4L2_MEMORY_MMAP? userptr already point to the right place.

File: drivers/video/video.c [hunk 21]
```diff
@@ -705,13 +710,26 @@ static void change_video_state(FAR video_mng_t    *vmng,
...
       if (container)
         {
-          start_capture(V4L2_BUF_TYPE_VIDEO_CAPTURE,
-                        vmng->video_inf.nr_fmt,
-                        vmng->video_inf.fmt,
-                        &vmng->video_inf.clip,
-                        &vmng->video_inf.frame_interval,
-                        container->buf.m.userptr,
-                        container->buf.length);
+          if (container->buf.memory == V4L2_MEMORY_MMAP)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:713]: why do you need check V4L2_MEMORY_MMAP? userptr already point to the right place.

File: drivers/video/video.c [hunk 22]
```diff
@@ -3331,8 +3444,18 @@ static int video_complete_capture(uint8_t  err_code, uint32_t datasize)
         }
       else
         {
-          g_video_data_ops->set_buf((FAR uint8_t *)container->buf.m.userptr,
-                                    container->buf.length);
+          if (container->buf.memory == V4L2_MEMORY_MMAP)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:3447]: why do you need check V4L2_MEMORY_MMAP? userptr already point to the right place.

File: drivers/video/video.c [hunk 23]
```diff
@@ -1800,7 +1901,7 @@ static int video_streamoff(FAR struct video_mng_s *vmng,
   irqstate_t           flags;
   int                  ret = OK;
 
-  if ((vmng == NULL) || (type == NULL))
+  if (vmng == NULL || type == NULL)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1904]: let's revert the change, I have a patch to update all place

File: drivers/video/video.c [hunk 24]
```diff
@@ -1757,7 +1858,7 @@ static int video_streamon(FAR struct video_mng_s *vmng,
   enum video_state_e   next_video_state;
   int                  ret = OK;
 
-  if ((vmng == NULL) || (type == NULL))
+  if (vmng == NULL || type == NULL)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1861]: let's revert the change, I have a patch to update all place

File: drivers/video/video.c [hunk 25]
```diff
@@ -1191,7 +1260,7 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
   enum video_state_e   next_video_state;
   irqstate_t           flags;
 
-  if ((vmng == NULL) || (buf == NULL))
+  if (vmng == NULL || buf == NULL)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1263]: let's revert the change, I have a patch to update all place

File: drivers/video/video.c [hunk 26]
```diff
@@ -1151,7 +1174,7 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
   FAR video_type_inf_t *type_inf;
   irqstate_t           flags;
 
-  if ((vmng == NULL) || (reqbufs == NULL))
+  if (vmng == NULL || reqbufs == NULL)
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1177]: let's revert the change, I have a patch to update all place

File: drivers/video/video.c [hunk 27]
```diff
@@ -1172,17 +1195,63 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
...
+      if (reqbufs->count > CONFIG_VIDEO_REQBUFS_COUNT_MAX)
+        {
+          reqbufs->count = CONFIG_VIDEO_REQBUFS_COUNT_MAX;
+        }
+
       video_framebuff_change_mode(&type_inf->bufinf, reqbufs->mode);
 
       ret = video_framebuff_realloc_container(&type_inf->bufinf,
                                               reqbufs->count);
+      if ((ret == OK) && (reqbufs->memory == V4L2_MEMORY_MMAP))
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1207]:
```suggestion
      if (ret == OK && reqbufs->memory == V4L2_MEMORY_MMAP)
```

File: drivers/video/video.c [hunk 28]
```diff
@@ -1172,17 +1195,63 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
...
   else
     {
+      if (reqbufs->count > CONFIG_VIDEO_REQBUFS_COUNT_MAX)
+        {
+          reqbufs->count = CONFIG_VIDEO_REQBUFS_COUNT_MAX;
+        }
+
       video_framebuff_change_mode(&type_inf->bufinf, reqbufs->mode);
 
       ret = video_framebuff_realloc_container(&type_inf->bufinf,
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1205]: let's remove the count check in video_framebuff_realloc_container and V4L2_REQBUFS_COUNT_MAX

File: drivers/video/video.c (hunk 17 above)
Comment (xiaoxiang781216) [reply drivers/video/video.c:226]: video_fmt_buf_size->get_bufsize to align with is_bufsize_sufficient

File: drivers/video/video.c [hunk 29]
```diff
@@ -940,6 +958,11 @@ static void cleanup_streamresources(FAR video_type_inf_t *type_inf)
   video_framebuff_uninit(&type_inf->bufinf);
   nxsem_destroy(&type_inf->wait_capture.dqbuf_wait_flg);
   nxmutex_destroy(&type_inf->lock_state);
+  if (type_inf->vmem_heap != NULL)
+    {
+      kmm_free(type_inf->vmem_heap);
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:963]:
```suggestion
      kumm_free(type_inf->vmem_heap);
```

File: drivers/video/video.c [hunk 30]
```diff
@@ -1172,17 +1195,63 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
...
+
       video_framebuff_change_mode(&type_inf->bufinf, reqbufs->mode);
 
       ret = video_framebuff_realloc_container(&type_inf->bufinf,
                                               reqbufs->count);
+      if ((ret == OK) && (reqbufs->memory == V4L2_MEMORY_MMAP))
+        {
+          if (type_inf->vmem_heap != NULL)
+            {
+              kmm_free(type_inf->vmem_heap);
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1211]:
```suggestion
              kumm_free(type_inf->vmem_heap);
```

File: drivers/video/video.c [hunk 31]
```diff
@@ -1172,17 +1195,63 @@ static int video_reqbufs(FAR struct video_mng_s         *vmng,
...
+            }
+
+          type_inf->vmem_heap = kmm_memalign(32, reqbufs->count *
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1214]:
```suggestion
          type_inf->vmem_heap = kumm_memalign(32, reqbufs->count *
```

File: drivers/video/video.c [hunk 32]
```diff
@@ -123,6 +126,7 @@ struct video_type_inf_s
   struct v4l2_rect     clip;
   struct v4l2_fract    frame_interval;
   video_framebuff_t    bufinf;
+  FAR void             *vmem_heap;   /* for V4L2_MEMORY_MMAP buffers */
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:129]:
```suggestion
  FAR uint8_t          *vmem_heap;   /* for V4L2_MEMORY_MMAP buffers */
```
help to compute the buffer location

File: drivers/video/video.c (hunk 20 above)
Comment (PeterBee97) [reply drivers/video/video.c:1309]: userptr in MMAP case is a union shared with offset, so it is not pointing to the right place?

File: drivers/video/video.c [hunk 33]
```diff
@@ -1243,6 +1302,7 @@ static int video_qbuf(FAR struct video_mng_s *vmng,
                             &type_inf->frame_interval,
                             container->buf.m.userptr,
                             container->buf.length);
+
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:1305]: revert the change

File: drivers/video/video.c [hunk 34]
```diff
@@ -62,6 +62,8 @@
 
 #define VIDEO_ID(x, y) (((x) << 16) | (y))
 
+#define MAX_VIDEO_HEAP_SIZE CONFIG_VIDEO_HEAP_SIZE
```
Comment (xiaoxiang781216) [inline drivers/video/video.c:65]: remove

File: drivers/video/video.c (hunk 28 above)
Comment (PeterBee97) [reply drivers/video/video.c:1184]: And also remove everything related to V4L2_REQBUFS_COUNT_MAX?

Comment (PeterBee97) [reply drivers/video/video.c:1184]: done

**Review by xiaoxiang781216 — APPROVED** [review]

File: drivers/video/video.c [hunk 35]
```diff
@@ -1522,6 +1584,22 @@ static int validate_frame_setting(enum v4l2_buf_type type,
   return g_video_data_ops->validate_frame_setting(nr_fmt, df, &di);
 }
 
+static size_t get_bufsize(video_format_t *vf)
```
Comment (pkarashchenko) [inline drivers/video/video.c:1587]: FAR
